diff -p -up validator/Makefile.in.DESTDIR validator/Makefile.in
--- validator/Makefile.in.DESTDIR	2011-08-01 16:09:04.342624270 -0300
+++ validator/Makefile.in	2011-08-01 16:11:17.016626500 -0300
@@ -4,7 +4,19 @@ LIBSUBDIRS=libsres libval libval_shim
 WINSUBDIRS=libsres libval apps 
 BUILD_ARCH = build
 
-INSTALLDIRS=$(prefix) $(exec_prefix) $(sysconfdir)/dnssec-tools $(bindir) $(sbindir) $(libdir) $(datadir) $(includedir) $(mandir) $(man1dir) $(man3dir) $(man5dir) $(man8dir)		
++INSTALLDIRS="$(DESTDIR)$(prefix) \
+	+    $(DESTDIR)$(exec_prefix) \
+	+    $(DESTDIR)$(sysconfdir)/dnssec-tools \
+	+    $(DESTDIR)$(bindir) \
+	+    $(DESTDIR)$(sbindir) \
+	+    $(DESTDIR)$(libdir) \
+	+    $(DESTDIR)$(datadir) \
+	+    $(DESTDIR)$(includedir) \
+	+    $(DESTDIR)$(mandir) \
+	+    $(DESTDIR)$(man1dir) \
+	+    $(DESTDIR)$(man3dir) \
+	+    $(DESTDIR)$(man5dir) \
+	+    $(DESTDIR)$(man8dir)"
 
 all: subdirmake
 
